pkg/storegateway/indexheader/stream_binary_reader.go

```diff
@@ -148,7 +174,87 @@ func newFileStreamBinaryReader(path string, postingOffsetsInMemSampling int, log
 		return nil, err
 	}
 
-	return r, nil
+	return r, err
+}
+
+// loadSparseFromDisk loads in sparse index-header from disk.
+func (r *StreamBinaryReader) loadSparseFromDisk(logger log.Logger, sparseHeadersPath string, sparseData []byte, postingOffsetsInMemSampling int) (err error) {
+	level.Debug(logger).Log("msg", "reading from index-header sparseHeaders file", "path", sparseHeadersPath)
+
+	sparseHeaders := &indexheaderpb.Sparse{}
+
+	gzipped := bytes.NewReader(sparseData)
+	gzipReader, err := gzip.NewReader(gzipped)
+	if err != nil {
+		return fmt.Errorf("failed to create index-header sparseHeaders reader: %w", err)
+	}
+
+	sparseData, err = io.ReadAll(gzipReader)
+	if err != nil {
+		return fmt.Errorf("failed to read index-header sparseHeaders: %w", err)
+	}
+
+	if err := sparseHeaders.Unmarshal(sparseData); err != nil {
+		return fmt.Errorf("failed to decode index-header sparseHeaders file: %w", err)
+	}
+
+	r.symbols, err = streamindex.NewSymbolsFromSparseHeader(r.factory, sparseHeaders.Symbols, r.indexVersion, int(r.toc.Symbols))
+	if err != nil {
+		return fmt.Errorf("cannot load symbols from sparse index-header: %w", err)
+	}
+
+	r.postingsOffsetTable, err = streamindex.NewPostingOffsetTableFromSparseHeader(r.factory, sparseHeaders.PostingsOffsetTable, int(r.toc.PostingsOffsetTable), postingOffsetsInMemSampling)
+	if err != nil {
+		return fmt.Errorf("cannot load postings offset table from sparse index-header: %w", err)
+	}
+
+	return nil
+}
+
+// loadSparseFromIndexHeader loads in symbols and postings offset table from the index-header and stores a sparse version.
+func (r *StreamBinaryReader) loadSparseFromIndexHeader(logger log.Logger, sparseHeadersPath string, cfg Config, indexLastPostingListEndBound uint64, postingOffsetsInMemSampling int) (err error) {
```

Comment:
1. This isn't loading the sparse version, it's just loading the regular version
2. This method doesn't need the sparse header path

```suggestion
func (r *StreamBinaryReader) loadFromIndexHeader(logger log.Logger, cfg Config, indexLastPostingListEndBound uint64, postingOffsetsInMemSampling int) (err error) {
```

Reply:
The version of the index-header saved to memory on a `StreamBinaryReader` is a sparse version. The `NewSymbol` and `NewPostingOffsetTable` methods automatically only sample 1/32. Technically, both `loadSparseFromIndexHeader` and `loadSparseFromDisk` are loading in the same thing, just from different places.

Reply:
Technically correct but still misleading. It's loading from the full index header file on disk, the "sparse" doesn't describe the in-memory representation - it describes the disk format. Please change it.